### PR TITLE
chore(core): remove Deref and DerefMut for Polynomial

### PR DESCRIPTION
**Pull request by IceTDrinker**
- add indexing and iteration primitives

### CODE REVIEW

File: tfhe/src/core_crypto/entities/polynomial.rs [hunk 1]
```diff
@@ -22,6 +24,104 @@ impl<T, C: ContainerMut<Element = T>> AsMut<[T]> for Polynomial<C> {
...
+        self.as_ref().index(index)
+    }
+}
+
+impl<C: ContainerMut> IndexMut<RangeToInclusive<usize>> for Polynomial<C> {
+    fn index_mut(&mut self, index: RangeToInclusive<usize>) -> &mut Self::Output {
+        self.as_mut().index_mut(index)
+    }
+}
+
```
Comment (tmontaigu) [inline tfhe/src/core_crypto/entities/polynomial.rs:124]: Maybe it is possible to shorten these with 

```rust
impl<C, I> Index<I> for Polynomial<C>
where C: Container + Index<I> {
    type Output = <C as Index<I>>::Output;

    fn index(&self, index: I) -> &Self::Output {
        self.as_ref().index(index)
    }
}
```

and Same for IndexMut

Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/polynomial.rs:124]: I'll check, but I tried a generic type name without bound it did not seem to work ok, maybe I wrote it wrong

Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/polynomial.rs:124]: actually, I was trying to go through the self.as_ref() that returns a slice and the indexing seemed to not work ok the first time around, now looks to be fine, must have been a bit tired Friday, updating

Comment (tmontaigu) [reply tfhe/src/core_crypto/entities/polynomial.rs]: Yeah if you use .as_ref()/.as_mut the bounds would have to be `for <'a> &'a [C::Element]: Index<I>`

Comment (IceTDrinker) [reply tfhe/src/core_crypto/entities/polynomial.rs]: looks like it was a bit easier than that, but this time I managed to get it working 🤷‍♂️

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**